### PR TITLE
Added new logic for template url parsing and tests

### PR DESCRIPTION
**Pull request by gargana**
## Overview

This adds better logic for URL parsing to locate nested templates

## Testing/Steps taken to ensure quality

Added tests and ran taskcat against some templates

### Notes

Still need to add the S3 fetching logic and tests for the same

## Testing Instructions

How to test this PR Start after checking out this branch (bulleted)

### CODE REVIEW

**Comment by avattathil** [comment]
@gargana I will have a PR in shortly that adds some added e2e tests (test run). Let me complete that before we merge this

File: taskcat/_cfn/template.py [hunk 1]
```diff
@@ -37,7 +37,7 @@ def __repr__(self):
 
     @property
     def s3_key(self):
-        suffix = str(self.template_path.relative_to(self.project_root).as_posix())
+        suffix = str(self.template_path.relative_to(self.project_root))
```
Comment (andrew-glenn) [inline taskcat/_cfn/template.py]: plz add this back in.

Comment (andrew-glenn) [reply taskcat/_cfn/template.py]: meaning, please rebase from master.

**Comment by gargana** [comment]
> See comment above.
> 
> Additionally, is it possible to remove a bunch of the bloat in these test templates? The additional resources aren't needed. There's huge value in focusing specifically on the resources that have a TemplateURL to parse - versus IAM Roles, etc etc.

I have resolved this now

File: taskcat/_cfn/stack_url_helper.py [hunk 2]
```diff
@@ -40,35 +40,33 @@ class StackURLHelper:
...
-            template_parameters = {}
-        else:
-            # print(template_parameters)
-            default_parameters = None
-            for parameter in template_parameters:
-                properties = template_parameters.get(parameter)
-                if "Default" in properties.keys():
-                    default_parameters.append(parameter)
+        if template_parameters:
+            self.template_parameters = template_parameters
```
Comment (andrew-glenn) [inline taskcat/_cfn/stack_url_helper.py]: The default value is none. 

Recommend:
```
if template_parameters:
  self.template_parameters = template_parameters
else self.template_parameters = {}
```

OR

```
template_parameters: dict = {}
```
in `__init__()`

File: taskcat/_cfn/stack_url_helper.py [hunk 3]
```diff
@@ -40,35 +40,33 @@ class StackURLHelper:
...
+
+        if parameter_values:
+            self.parameter_values = parameter_values
 
-            if default_parameters:
-                self.SUBSTITUTION.update(default_parameters)
+        default_parameters = None
+        for parameter in self.template_parameters:
```
Comment (andrew-glenn) [inline taskcat/_cfn/stack_url_helper.py]: If this is `None`, you'll get an exception here. See note above

File: tests/data/stackurlhelper/test.json [hunk 4]
```diff
@@ -3,11 +3,15 @@
...
+                "child_template": "{'Fn::Sub': ['${QuickStartS3URL}/${QSS3BucketName}/${QSS3KeyPrefix}/templates/test_mappings_child.yml', {'QuickStartS3URL': {'Fn::FindInMap': [ 'AWSInfoRegionMap', {'Ref': 'AWS::Region'}, 'QuickStartS3URL']}}]}",
+                "parameter_values": {
+                    "QSS3BucketName": "aws-quickstart",
+                    "QSS3KeyPrefix": "QSS3Prefix/"
+                }
             },
             "output": {
                 "url_paths": [
-                    "/aws-quickstart/QSS3KeyPrefix//templates/test_mappings_child.yml"
+                    "/aws-quickstart/quickstart-aws-eks//templates/test_mappings_child.yml"
```
Comment (andrew-glenn) [inline tests/data/stackurlhelper/test.json]: Double-slash? Is that intentional?

File: tests/test_stack_url_helper.py [hunk 5]
```diff
@@ -24,21 +24,37 @@ def _load_template(template_path):
...
 
-        helper = StackURLHelper()
-        for test in self.tests:
+        for test in self.testers:
+            helper = StackURLHelper()
             cfn = self._load_template(test["input"]["master_template"])
             helper.mappings = cfn.get("Mappings")
+            helper.template_parameters = cfn.get("Parameters")
+
+            # Setup default parameters
```
Comment (andrew-glenn) [inline tests/test_stack_url_helper.py]: Same `default_parameters` comment from above.

File: tests/data/stackurlhelper/test.json (hunk 4 above)
Comment (gargana) [reply tests/data/stackurlhelper/test.json]: Yes that is.

File: taskcat/_cfn/stack_url_helper.py (hunk 3 above)
Comment (gargana) [reply taskcat/_cfn/stack_url_helper.py]: This is fixed in the latest commit that was pushed

File: taskcat/_cfn/stack_url_helper.py (hunk 2 above)
Comment (gargana) [reply taskcat/_cfn/stack_url_helper.py]: resolved

**Comment by avattathil** [comment]
Reverting this PR. New logic does not account for unhanded substitutions 
See here https://gist.github.com/avattathil/66878f92ed435fb6b68fb534295c6f3e

**Comment by f4bi4n89** [comment]
Will this new logic also includes the `update-ami`? For example in my case, I use nested stacks and in one of these is the EC2 stuff with the `AWSAMIRegionMap`. But when I use  `update-ami`  on my main template, it ignores the EC2 template and doesn't update the AMI IDs.  

Is there another way to run the update on such files, which I'm missing? Because of the missing  `-i INPUT_FILE`  in  `update-ami` , I can't call it with a separate config file which could include the ec2 template.

thanks in advanced!

**Comment by andrew-glenn** [comment]
@f4bi4n89 

On the surface, it sounds like this will address what you're describing. FYI we reverted this PR as we found unexpected behavior. The next iteration of this PR is #499

**Pull request merged**